### ISSUE TITLE
Windows 10: Add code to obey "Dark mode" theme, if that is even possible

### ISSUE THREAD

**Issue by joncampbell123** [joncampbell123/dosbox-x#1607]
On a desktop with a dark mode theme, DOSBox-X like most desktop applications still appears with a white title bar and menu bar.

Apparently whatever API or magic is involved to let the command prompt and Explorer support it is undocumented.

There is a Github project to hack around with undocumented ordinals, but that is not a good idea.

At some point when Microsoft finally documents how to enable DOSBox-X to obey the dark mode theme, it should be added to the code base in Windows builds. Using LoadModule and GetProcAddress of course so downlevel Windows can still run DOSBox-X.

**Comment by sikthehedgehog** [comment]
Looking up and it does seem that Microsoft is intentionally trying to keep it an UWP-only feature… though they didn't do a great job in that sense.

I found this, they mention the registry method but some answers also have less invasive solutions (possibly more futureproof), not sure how many of those work but may be worth looking into:
https://stackoverflow.com/questions/51334674/how-to-detect-windows-10-light-dark-mode-in-win32-application

Also, Microsoft is well aware that programmers will get their way and I imagine they won't want to risk breaking it once several applications have started to rely on a given quirk :P Also the idea of trying to guess the mode by the default colors could be worth checking in the worst case >_> (though I'd try to compare colors rather than looking for specific ones, which may make it work in more scenarios)

Looks like dark mode works by letting the application pick its own colors though (welcome to the web-like era of development), so if you're trying to stick to the "default" ones then that may backfire.

**Comment by ThePillenwerfer** [comment]
DOSBox-X ignores the OS theme completely on Linux.  The menu labels and the menus themselves are always light grey text in the DOSBox-X font on dark grey no matter what theme is set.

I don't regard it as a problem as when I've got a program set-up properly in DOSBox-X I'll odds-on actually use it in fullscreen mode so won't see the menus.  Obviously it'd be nice if it matched everything else but making it do so would be a low priority.

**Comment by NightmareJoker2** [comment]
> Apparently whatever API or magic is involved to let the command prompt and Explorer support it is undocumented.

It should still work the same way it always has, if you aren't using the default OS controls, you grab the appropriate color using [GetSysColor](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getsyscolor) and the fonts with [SystemParametersInfoA ](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-systemparametersinfoa) and set it on your custom control. But apparently **it doesn't**, and the new "dark mode" doesn't change any of the legacy "theme" colors (previously changeable in the Color and Appearance dialogs in the control panel) which have been removed in Windows 8 or 8.1. The settings do however still work, and you can still change them with something like [Classic Color Panel](https://www.wintools.info/index.php/classic-color-panel) and [Advanced System Font Changer](https://www.wintools.info/index.php/advanced-system-font-changer) (or directly in the registry as RGB values stored at *HKCU/Control Panel/Colors*). Because the settings UIs are now gone, the Win32 API broadcasts to tell application windows to update the interface colors now also does nothing and the colors are now only updated when you log on, meaning you need to sign out and sign back in or restart your computer (if you changed the colors of the default user to affect the login screen, if you're using the classic kind) to have them take effect.
You will still get a [WM_WININICHANGE](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-wininichange), and even dark mode uses this broadcast with `lParam` set to `ImmersiveColorSet` to notify of a theme change, after which you can then check the color, theme, and light/dark mode settings and apply them to the control.
Basically, though, the issue is Windows ignoring legacy app support with dark mode, so far, and dark mode is only a binary setting at *HKCU\Software\Microsoft\Windows\CurrentVersion\Themes\Personalize\AppsUseLightTheme* which doesn't indicate at all what dark means and what light means (with an API that returns colors only returning white or black), and the expectation that an application decides what to do. This is evident also with applications that ship with Windows, such as Windows Explorer's settings, the classic control panel, task manager, and pretty much everything that is not a modern UWP app not respecting dark mode at all. You can fix this by changing the colors with the Classic Color Changer app, I linked to above. If the menu bar is still the default Windows gray after doing that, though, there is indeed a bug in the menu control.

**Comment by joncampbell123** [comment]
In other words, Microsoft implemented "dark mode" badly in a way that is going to bite them hard in the future should they add any additional modes, while failing to take advantage of the API that has literally existed since Windows 1.0 to define window colors.

Look, Windows 1.0 has dark mode too!
![win100_000 avi](https://user-images.githubusercontent.com/6245486/83812342-3ab96d00-a670-11ea-9d60-aa49c5bd24e2.gif)

**Comment by NightmareJoker2** [comment]
> In other words, Microsoft implemented "dark mode" [...] failing to take advantage of the API that has literally existed since Windows 1.0 to define window colors.

Yes, pretty much. Part of the rationale being, that title bars, button faces, progress bars and window borders are affected by [Visual Styles](https://docs.microsoft.com/en-us/windows/win32/controls/using-visual-styles) introduced in Windows XP, the user creation of which Microsoft never supported (even though it'd work just fine with a little patch to *uxtheme.dll*, something which even software like Stardock's [WindowBlinds](https://www.stardock.com/products/windowblinds/) then took advantage of from XP onwards), modern Windows no longer offering the classic style, where coloring works properly (though that is only half true, since visual styles rely on a theme file that defines the colors for those elements a visual style does not or cannot offer graphics for, and themes with visual styles being able to offer color variations, much like the three defaults in Windows XP, which could be made dark mode aware, *and* the high contrast themes for accessibility still making it partially available, color settings included), and most users not ever bothering to change the default colors anymore, according to their Insider/Customer Improvement Program telemetry (where this is then presumably too much work to maintain the feature, but I don't know).
